File: backend/src/test/java/org/globe42/web/users/UserControllerTest.java
package org.globe42.web.users;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

import java.util.Collections;
import java.util.List;
import java.util.Optional;

import org.globe42.dao.NoteDao;
import org.globe42.dao.SpentTimeDao;
import org.globe42.dao.TaskDao;
import org.globe42.dao.UserDao;
import org.globe42.domain.User;
import org.globe42.test.BaseTest;
import org.globe42.web.exception.BadRequestException;
import org.globe42.web.exception.NotFoundException;
import org.globe42.web.security.CurrentUser;
import org.globe42.web.security.PasswordDigester;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.InjectMocks;
import org.mockito.Mock;

/**
 * Unit tests for {@link UserController}
 * @author JB Nizet
 */
public class UserControllerTest extends BaseTest {
    @Mock
    private CurrentUser mockCurrentUser;

    @Mock
    private UserDao mockUserDao;

    @Mock
    private TaskDao mockTaskDao;

    @Mock
    private NoteDao mockNoteDao;

    @Mock
    private SpentTimeDao mockSpentTimeDao;

    @Mock
    private PasswordGenerator mockPasswordGenerator;

    @Mock
    private PasswordDigester mockPasswordDigester;

    @InjectMocks
    private UserController controller;

    @Captor
    private ArgumentCaptor<User> userCaptor;

    private Long userId = 42L;

    @BeforeEach
    public void prepare() {
        when(mockCurrentUser.getUserId()).thenReturn(userId);
    }

    @Test
    public void shouldGetCurrentUser() {
        User user = createUser(userId);
        when(mockUserDao.findNotDeletedById(userId)).thenReturn(Optional.of(user));

        CurrentUserDTO result = controller.getCurrentUser();
        assertThat(result.getId()).isEqualTo(user.getId());
        assertThat(result.getLogin()).isEqualTo(user.getLogin());
    }

    @Test
    public void shouldThrowWhenGettingCurrentUserIfNotFound() {
        when(mockUserDao.findNotDeletedById(userId)).thenReturn(Optional.empty());

        assertThatExceptionOfType(NotFoundException.class).isThrownBy(() -> controller.getCurrentUser());
    }

    @Test
    public void shouldChangePasswordOfCurrentUser() {
        User user = createUser(userId);
        when(mockUserDao.findNotDeletedById(userId)).thenReturn(Optional.of(user));

        ChangePasswordCommandDTO command = new ChangePasswordCommandDTO("newPassword");
        when(mockPasswordDigester.hash(command.getNewPassword())).thenReturn("hashed");

        controller.changePassword(command);

        assertThat(user.getPassword()).isEqualTo("hashed");
    }

    @Test
    public void shouldList() {
        User user = createUser(userId);
        when(mockUserDao.findNotDeleted()).thenReturn(Collections.singletonList(user));

        List<UserDTO> result = controller.list();
        assertThat(result).hasSize(1);
        assertThat(result.get(0).getId()).isEqualTo(user.getId());
        assertThat(result.get(0).getLogin()).isEqualTo(user.getLogin());
    }

    @Test
    public void shouldGet() {
        User user = createUser(userId);
        when(mockUserDao.findNotDeletedById(userId)).thenReturn(Optional.of(user));

        UserDTO result = controller.get(userId);
        assertThat(result.getId()).isEqualTo(user.getId());
    }

    @Test
    public void shouldThrowWhenGettingIfNotFound() {
        when(mockUserDao.findNotDeletedById(userId)).thenReturn(Optional.empty());
        assertThatExceptionOfType(NotFoundException.class).isThrownBy(() -> controller.get(userId));
    }

    @Test
    public void shouldCreate() {
        UserCommandDTO command = new UserCommandDTO("test", true);

        when(mockPasswordGenerator.generatePassword()).thenReturn("password");
        when(mockPasswordDigester.hash("password")).thenReturn("hashed");

        UserWithPasswordDTO result = controller.create(command);
        assertThat(result.getGeneratedPassword()).isEqualTo("password");
        assertThat(result.getUser().getLogin()).isEqualTo(command.getLogin());
        assertThat(result.getUser().isAdmin()).isEqualTo(command.isAdmin());

        verify(mockUserDao).save(userCaptor.capture());
        assertThat(userCaptor.getValue().getPassword()).isEqualTo("hashed");
    }

    @Test
    public void shouldThrowWhenCreatingWithExistingLogin() {
        UserCommandDTO command = new UserCommandDTO("test", false);

        when(mockUserDao.existsByLogin(command.getLogin())).thenReturn(true);

        assertThatExceptionOfType(BadRequestException.class).isThrownBy(() -> controller.create(command));
    }

    @Test
    public void shouldUpdate() {
        UserCommandDTO command = new UserCommandDTO("test", true);
        User user = createUser(userId);
        when(mockUserDao.findNotDeletedById(userId)).thenReturn(Optional.of(user));

        controller.update(userId, command);
        assertThat(user.getLogin()).isEqualTo(command.getLogin());
        assertThat(user.isAdmin()).isEqualTo(command.isAdmin());
    }

    @Test
    public void shouldThrowWhenUpdatingWithExistingLogin() {
        UserCommandDTO command = new UserCommandDTO("test", false);
        User user = createUser(userId);
        when(mockUserDao.findNotDeletedById(userId)).thenReturn(Optional.of(user));

        when(mockUserDao.findNotDeletedByLogin(command.getLogin())).thenReturn(Optional.of(createUser(4567L)));

        assertThatExceptionOfType(BadRequestException.class).isThrownBy(() -> controller.update(userId, command));
    }

    @Test
    public void shouldNotThrowWhenUpdatingWithSameLogin() {
        User user = createUser(userId);
        UserCommandDTO command = new UserCommandDTO(user.getLogin(), false);
        when(mockUserDao.findNotDeletedById(userId)).thenReturn(Optional.of(user));

        when(mockUserDao.findNotDeletedByLogin(command.getLogin())).thenReturn(Optional.of(user));

        controller.update(userId, command);
    }

    @Test
    public void shouldDelete() {
        User user = createUser(userId);
        when(mockUserDao.findNotDeletedById(userId)).thenReturn(Optional.of(user));

        controller.delete(userId);

        assertThat(user.isDeleted()).isTrue();
    }

    @Test
    public void shouldNotDoAnythingWhenDeletingUnexistingUser() {
        when(mockUserDao.findNotDeletedById(userId)).thenReturn(Optional.empty());

        controller.delete(userId);

        verify(mockUserDao, never()).delete(any(User.class));
    }

    @Test
    public void shouldResetPassword() {
        User user = createUser(userId);
        when(mockUserDao.findNotDeletedById(userId)).thenReturn(Optional.of(user));

        when(mockPasswordGenerator.generatePassword()).thenReturn("password");
        when(mockPasswordDigester.hash("password")).thenReturn("hashed");
        UserWithPasswordDTO result = controller.resetPassword(userId);

        assertThat(user.getPassword()).isEqualTo("hashed");
        assertThat(result.getGeneratedPassword()).isEqualTo("password");
    }

    public static User createUser(Long id) {
        User user = new User(id);
        user.setLogin("JB");
        user.setPassword("hashedPassword");
        return user;
    }
}

File: backend/src/test/kotlin/org/globe42/web/users/UserControllerTest.kt
package org.globe42.web.users

import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatExceptionOfType
import org.globe42.dao.UserDao
import org.globe42.domain.User
import org.globe42.web.exception.BadRequestException
import org.globe42.web.exception.NotFoundException
import org.globe42.web.security.CurrentUser
import org.globe42.web.security.PasswordDigester
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

fun createUser(id: Long): User {
    return User(id).apply {
        login = "JB"
        password = "hashedPassword"
        email = "jb@foo.com"
        taskAssignmentEmailNotificationEnabled = false
    }
}

/**
 * Unit tests for [UserController]
 * @author JB Nizet
 */
class UserControllerTest {
    private val mockCurrentUser = mockk<CurrentUser>()

    private val mockUserDao = mockk<UserDao>()

    private val mockPasswordGenerator = mockk<PasswordGenerator>()

    private val mockPasswordDigester = mockk<PasswordDigester>()

    private val controller = UserController(mockCurrentUser, mockUserDao, mockPasswordGenerator, mockPasswordDigester)

    private val userId = 42L

    @BeforeEach
    fun prepare() {
        every { mockCurrentUser.userId } returns userId
    }

    @Test
    fun `should get current user`() {
        val user = createUser(userId)
        every { mockUserDao.findNotDeletedById(userId) } returns user

        val (id, login) = controller.getCurrentUser()
        assertThat(id).isEqualTo(user.id)
        assertThat(login).isEqualTo(user.login)
    }

    @Test
    fun `should throw when getting current user if not found`() {
        every { mockUserDao.findNotDeletedById(userId) } returns null

        assertThatExceptionOfType(NotFoundException::class.java).isThrownBy { controller.getCurrentUser() }
    }

    @Test
    fun `should change password of current user`() {
        val user = createUser(userId)
        every { mockUserDao.findNotDeletedById(userId) } returns user

        val command = ChangePasswordCommandDTO("newPassword")
        every { mockPasswordDigester.hash(command.newPassword) } returns "hashed"

        controller.changePassword(command)

        assertThat(user.password).isEqualTo("hashed")
    }

    @Test
    fun `should list`() {
        val user = createUser(userId)
        every { mockUserDao.findNotDeleted() } returns listOf(user)

        val result = controller.list()
        assertThat(result).hasSize(1)
        assertThat(result[0].id).isEqualTo(user.id)
        assertThat(result[0].login).isEqualTo(user.login)
    }

    @Test
    fun `should get`() {
        val user = createUser(userId)
        every { mockUserDao.findNotDeletedById(userId) } returns user

        val (id) = controller.get(userId)
        assertThat(id).isEqualTo(user.id)
    }

    @Test
    fun `should throw when getting if not found`() {
        every { mockUserDao.findNotDeletedById(userId) } returns null
        assertThatExceptionOfType(NotFoundException::class.java).isThrownBy { controller.get(userId) }
    }

    @Test
    fun `should create`() {
        val command = UserCommandDTO("test", true)

        every { mockUserDao.existsByLogin(command.login) } returns false
        every { mockPasswordGenerator.generatePassword() } returns "password"
        every { mockPasswordDigester.hash("password") } returns "hashed"
        every { mockUserDao.save(any<User>()) } answers { arg<User>(0).apply { id = 42L } }

        val (user, generatedPassword) = controller.create(command)
        assertThat(generatedPassword).isEqualTo("password")
        assertThat(user.login).isEqualTo(command.login)
        assertThat(user.admin).isEqualTo(command.admin)

        verify {
            mockUserDao.save(withArg<User> {
                assertThat(it.password).isEqualTo("hashed")
            })
        }
    }

    @Test
    fun `should throw when creating with existing login`() {
        val command = UserCommandDTO("test", false)

        every { mockUserDao.existsByLogin(command.login) } returns true

        assertThatExceptionOfType(BadRequestException::class.java).isThrownBy { controller.create(command) }
    }

    @Test
    fun `should update`() {
        val command = UserCommandDTO("test", true)
        val user = createUser(userId)

        every { mockUserDao.findNotDeletedByLogin(command.login) } returns null
        every { mockUserDao.findNotDeletedById(userId) } returns user

        controller.update(userId, command)
        assertThat(user.login).isEqualTo(command.login)
        assertThat(user.admin).isEqualTo(command.admin)
    }

    @Test
    fun `should throw when updating with existing login`() {
        val command = UserCommandDTO("test", false)
        val user = createUser(userId)
        every { mockUserDao.findNotDeletedById(userId) } returns user

        every { mockUserDao.findNotDeletedByLogin(command.login) } returns createUser(4567L)

        assertThatExceptionOfType(BadRequestException::class.java).isThrownBy { controller.update(userId, command) }
    }

    @Test
    fun `should not throw when updating with same login`() {
        val user = createUser(userId)
        val command = UserCommandDTO(user.login, false)
        every { mockUserDao.findNotDeletedById(userId) } returns user

        every { mockUserDao.findNotDeletedByLogin(command.login) } returns user

        controller.update(userId, command)
    }

    @Test
    fun `should delete`() {
        val user = createUser(userId)
        every { mockUserDao.findNotDeletedById(userId) } returns user

        controller.delete(userId)

        assertThat(user.deleted).isTrue()
    }

    @Test
    fun `should not do anything when deleting unexisting user`() {
        every { mockUserDao.findNotDeletedById(userId) } returns null

        controller.delete(userId)

        verify(inverse = true) { mockUserDao.delete(any()) }
    }

    @Test
    fun `should reset password`() {
        val user = createUser(userId)
        every { mockUserDao.findNotDeletedById(userId) } returns user

        every { mockPasswordGenerator.generatePassword() } returns "password"
        every { mockPasswordDigester.hash("password") } returns "hashed"
        val (_, generatedPassword) = controller.resetPassword(userId)

        assertThat(user.password).isEqualTo("hashed")
        assertThat(generatedPassword).isEqualTo("password")
    }
}
